x/liquidity/keeper/migration.go

```diff
@@ -1,22 +1,169 @@
+	// the corresponding amount of respective reserve coins back to their accounts.
+	// Lastly, burn pool coins that are withdrawn.
+	accMap := map[string]authtypes.AccountI{}
+	accList := []string{}
+	errorCount := 0
+	bankKeeper.IterateAllBalances(ctx, func(address sdk.AccAddress, coin sdk.Coin) (stop bool) {
+		if strings.HasPrefix(coin.Denom, "pool") {
+			addrStr := address.String()
+			pool := poolByPoolCoinDenom[coin.Denom]
+			if _, ok := accMap[addrStr]; !ok {
+				accMap[addrStr] = accountKeeper.GetAccount(ctx, address)
+				accList = append(accList, addrStr)
+			}
+
+			// Skip pool reserve accounts
+			acc := accMap[addrStr]
+			if acc.GetSequence() != 0 || acc.GetPubKey() != nil {
+				if res, err := keeper.WithdrawWithinBatch(ctx, &liquiditytypes.MsgWithdrawWithinBatch{
+					WithdrawerAddress: addrStr,
+					PoolId:            pool.GetId(),
+					PoolCoin:          coin,
+				}); !res.Succeeded || err != nil {
+					errorCount++
+				}
+			}
+		}
+		return false
+	})
+
+	// Execute batch manually
+	keeper.ExecutePoolBatches(ctx)
+	keeper.DeleteAndInitPoolBatches(ctx)
+
+	// iterating and withdraw again if there is any pool coin left due to decimal error by too small pool coin value
+	if errorCount > 0 {
+		for _, address := range accList {
+			acc := accMap[address]
+			balances := keeper.bankKeeper.GetAllBalances(ctx, acc.GetAddress())
+			for _, coin := range balances {
+				if strings.HasPrefix(coin.Denom, "pool") {
+					// Skip pool reserve accounts
+					if acc.GetSequence() != 0 || acc.GetPubKey() != nil {
+						if _, err := keeper.WithdrawWithinBatch(ctx, &liquiditytypes.MsgWithdrawWithinBatch{
+							WithdrawerAddress: address,
+							PoolId:            poolByPoolCoinDenom[coin.Denom].GetId(),
+							PoolCoin:          coin,
+						}); err != nil {
+							logger.Debug(
+								"failed force withdrawal",
+								"withdrawer", address,
+								"poolcoin", coin,
+								"error", err,
+							)
+						}
+					}
+				}
+			}
+		}
+
+		// Execute batch manually
+		keeper.ExecutePoolBatches(ctx)
+		keeper.DeleteAndInitPoolBatches(ctx)
+	}
+
+	if len(keeper.GetAllDepositMsgStates(ctx)) > 0 {
+		return fmt.Errorf("deposit msg states must be empty for migration")
+	}
+	if len(keeper.GetAllSwapMsgStates(ctx)) > 0 {
+		return fmt.Errorf("swap msg states must be empty for migration")
+	}
+	if len(keeper.GetAllWithdrawMsgStates(ctx)) > 0 {
+		return fmt.Errorf("withdraw msg states must be empty for migration")
+	}
+
+	// Fund remaining reserve balance to community pool
+	for _, pool := range keeper.GetAllPools(ctx) {
+		reserveAcc := pool.GetReserveAccount()
+		balances := keeper.bankKeeper.GetAllBalances(ctx, reserveAcc)
+		if balances.IsZero() {
+			continue
+		}
+		err := keeper.distrKeeper.FundCommunityPool(ctx, balances, reserveAcc)
+		if err != nil {
+			logger.Debug(
+				"failed fund community pool",
+				"pool id", pool.Id,
+				"error", err,
+			)
```

Comment:
if `FundCommunityPool` failed, no return err ?

Reply:
[FundCommunityPool](https://github.com/cosmos/cosmos-sdk/blob/v0.45.16-ics/x/distribution/keeper/keeper.go#L152-L166) is very 

Yes, The [FundCommunityPool](https://github.com/cosmos/cosmos-sdk/blob/v0.45.16-ics/x/distribution/keeper/keeper.go#L152-L166) is a very 
simple function that is all about Send, Get, Add, and Set. In general, it is difficult to reproduce the failure and the probability of failure is close to zero, so error handling was not performed. Even if all of the previous logic succeeded and failed at the here, rather than reverting the entire migration by returning the error, it can be solved by re-calling only the FundCommunityPool function in the subsequent upgrade, so I don't think it's necessary to revert.